### PR TITLE
Small improvements for Product and SyncPlan entities

### PR DESCRIPTION
**Pull request by abalakh**
For Product, added ability to sync all repos associated with it.
For SyncPlan, added ability to use `update()`.
(Need both changes for upcoming robottelo tests)

### CODE REVIEW

**Comment by abalakh** [comment]
Added some unit tests :)
Btw, @Ichimonji10 , is there any way to run coveralls tests locally? Just to save time and not to update PR only after check has failed.

**Comment by Ichimonji10** [comment]
@abalakh Take a look at the makefile. (I could just tell you the answer, but... :smile:)

**Comment by abalakh** [comment]
@Ichimonji10 i can see only `test-coverage` there. `test-coverage` shown 'OK' to me, but coveralls said that coverage decreased. That's why i'm asking :)

File: nailgun/entities.py [hunk 1]
```diff
@@ -3541,6 +3555,13 @@ def remove_products(self, product_ids, synchronous=True):
         )
         return _handle_response(response, self._server_config, synchronous)
 
+    def update_payload(self, fields=None):
+        """Convert ``sync_date`` to a string if datetime object provided."""
+        data = super(SyncPlan, self).update_payload()
+        if 'sync_date' in data and isinstance(data['sync_date'], datetime):
```
Comment (elyezer) [inline nailgun/entities.py]: `if isinstance (data.get('sync_date'), datetime):`?

File: tests/test_entities.py [hunk 2]
```diff
@@ -1208,6 +1230,22 @@ def test_repository_sets_available_repositories(self):
         self.assertEqual(handler.call_count, 1)
         self.assertEqual(handler.return_value['results'], response)
 
+    def test_synchronize(self):
+        """Call
```
Comment (elyezer) [inline tests/test_entities.py]: This docstring does not fit oneline?

File: tests/test_entities.py [hunk 3]
```diff
@@ -1208,6 +1230,22 @@ def test_repository_sets_available_repositories(self):
...
+        with mock.patch.object(client, 'post') as client_post:
+            with mock.patch.object(
+                entities,
+                '_handle_response',
+                return_value={'results': gen_integer()},  # not realistic
+            ) as handler:
+                response = self.product.synchronize()
+        self.assertEqual(client_post.call_count, 1)
+        self.assertEqual(handler.call_count, 1)
+        self.assertEqual(handler.return_value, response)
```
Comment (elyezer) [inline tests/test_entities.py]: This line seems to so useful as we are asserting that the mock library is doing its job. Maybe we can just drop this.

**Comment by elyezer** [comment]
ACK, pending not blocker comments.

File: nailgun/entities.py (hunk 1 above)
Comment (abalakh) [inline nailgun/entities.py]: :+1:

**Comment by Ichimonji10** [comment]
Could you please apply the diff below?

* Rename `synchronize` to `sync` so that the method name matches the underlying path.
* Rename several variables and tweak calls to `subTest()` for clarity's sake.
* Make a class handle just dicts instead of complete entity objects, thus reducing redundancy. The test code will become more compact and even less redundant after the `SyncPlan` class is updated with some changes introduced in Sat 6.1.1.
* Rewrite a docstring in the present tense. Make it possible to say "When this method is called, it will <insert docstring here>."

```diff
diff --git a/nailgun/entities.py b/nailgun/entities.py
index 37ce47e..8bc9e6a 100644
--- a/nailgun/entities.py
+++ b/nailgun/entities.py
@@ -2681,6 +2681,8 @@ class Product(
             /products/<product_id>/repository_sets/<id>/enable
         repository_sets/<id>/disable
             /products/<product_id>/repository_sets/<id>/disable
+        sync
+            /products/<product_id>/sync
 
         ``super`` is called otherwise.
 
@@ -2835,8 +2837,8 @@ class Product(
         )
         return _handle_response(response, self._server_config)['results']
 
-    def synchronize(self):
-        """Synchronizes all repos in a product"""
+    def sync(self):
+        """Synchronize :class:`repositories <Repository>` in this product."""
         response = client.post(
             self.path('sync'),
             {},
diff --git a/tests/test_entities.py b/tests/test_entities.py
index 4bd14ee..ddf4ad3 100644
--- a/tests/test_entities.py
+++ b/tests/test_entities.py
@@ -886,7 +886,7 @@ class UpdatePayloadTestCase(TestCase):
 
     def test_generic(self):
         """Instantiate a variety of entities and call ``update_payload``."""
-        class_response = [
+        entities_payloads = [
             (entities.ConfigTemplate, {'config_template': {}}),
             (entities.Domain, {'domain': {}}),
             (entities.Host, {'host': {}}),
@@ -895,32 +895,33 @@ class UpdatePayloadTestCase(TestCase):
             (entities.Organization, {'organization': {}}),
             (entities.User, {'user': {}}),
         ]
-        for entity, response in class_response:
-            with self.subTest():
-                self.assertEqual(
-                    entity(self.cfg).update_payload(),
-                    response
-                )
+        for entity, payload in entities_payloads:
+            with self.subTest((entity, payload)):
+                self.assertEqual(entity(self.cfg).update_payload(), payload)
 
     def test_syncplan_sync_date(self):
         """Test ``update_payload`` for different syncplan sync_date formats."""
         date_string = '2015-07-20 20:54:38'
         date_datetime = datetime.strptime(date_string, '%Y-%m-%d %H:%M:%S')
-        syncplan_response = [
-            (entities.SyncPlan(self.cfg, organization=1),
-             {'organization_id': 1}),
-            (entities.SyncPlan(self.cfg, organization=1,
-                               sync_date=date_string),
-             {'organization_id': 1, 'sync_date': date_string}),
-            (entities.SyncPlan(self.cfg, organization=1,
-                               sync_date=date_datetime),
-             {'organization_id': 1, 'sync_date': date_string}),
+        kwargs_responses = [
+            (
+                {'organization': 1},
+                {'organization_id': 1},
+            ),
+            (
+                {'organization': 1, 'sync_date': date_string},
+                {'organization_id': 1, 'sync_date': date_string},
+            ),
+            (
+                {'organization': 1, 'sync_date': date_datetime},
+                {'organization_id': 1, 'sync_date': date_string},
+            ),
         ]
-        for syncplan, response in syncplan_response:
-            with self.subTest():
+        for kwargs, payload in kwargs_responses:
+            with self.subTest((kwargs, payload)):
                 self.assertEqual(
-                    syncplan.update_payload(),
-                    response
+                    entities.SyncPlan(self.cfg, **kwargs).update_payload(),
+                    payload,
                 )
 
 
@@ -1231,14 +1232,14 @@ class ProductTestCase(TestCase):
         self.assertEqual(handler.return_value['results'], response)
 
     def test_synchronize(self):
-        """Call :meth:`nailgun.entities.Product.synchronize`."""
+        """Call :meth:`nailgun.entities.Product.sync`."""
         with mock.patch.object(client, 'post') as client_post:
             with mock.patch.object(
                 entities,
                 '_handle_response',
                 return_value={'results': gen_integer()},  # not realistic
             ) as handler:
-                self.product.synchronize()
+                self.product.sync()
         self.assertEqual(client_post.call_count, 1)
         self.assertEqual(handler.call_count, 1)
```

**Comment by Ichimonji10** [comment]
Thank *you*. ACK.

**Comment by Ichimonji10** [comment]
@elyezer Do you want to take a second look?

**Comment by sthirugn** [comment]
ACK

**Comment by omaciel** [comment]
@sthirugn @Ichimonji10 I got a message from @elyezer and he said you should not block on him to get this merged.

**Comment by sthirugn** [comment]
@omaciel ty

**Comment by elyezer** [comment]
:+1:

**Pull request merged**